File: advection/advection_mpi.c
/* Daniel R. Reynolds
   SMU Mathematics
   Math 4370/6370
   11 May 2017 */

/* Inclusions */
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <math.h>
#include "advection_mpi.h"
#include "mpi.h"


/* Example routine to evolve the first-order 2D wave equations in time. */
int main(int argc, char* argv[]) {

  /* declarations */
  int nx, ny, nt, i, j, it, ierr, ibuff[3];
  double tstop, c, dtoutput;
  double v1_N, v1_S, v1_E, v1_W, v2_E, v2_W, v3_N, v3_S, rbuff[3];

  /* intialize MPI */
  parallel_decomp p2d;
  create_parallel_decomp(&p2d);
  
  ierr = MPI_Init(&argc, &argv);
  check_err(ierr, MPI_COMM_WORLD, "MPI_Init");

  ierr = MPI_Comm_size(MPI_COMM_WORLD, &(p2d.numprocs));
  check_err(ierr, MPI_COMM_WORLD, "MPI_Comm_size");

  ierr = MPI_Comm_rank(MPI_COMM_WORLD, &(p2d.myid));
  check_err(ierr, MPI_COMM_WORLD, "MPI_Comm_size");

  double stime = MPI_Wtime();
  /* read problem parameters from input file (should be in this order):
        nx - number of nodes in x-direction
	ny - number of nodes in y-direction
	nt - number of time steps
	tstop - final time (will stop at nt or stop, whichever is 1st)
	c - wave speed
	dtoutput - time frequency for outputting solutions */
  if (p2d.myid == 0) {
    FILE *FID = fopen("input.txt","r");
    i = fscanf(FID," &inputs\n");
    i = fscanf(FID,"  nx = %i,\n", &nx);
    i = fscanf(FID,"  ny = %i,\n", &ny);
    i = fscanf(FID,"  nt = %i,\n", &nt);
    i = fscanf(FID,"  tstop = %lf,\n", &tstop);
    i = fscanf(FID,"  c = %lf,\n", &c);
    i = fscanf(FID,"  dtoutput = %lf,\n", &dtoutput);
    fclose(FID);

    printf("\nRunning wave problem:\n");
    printf("  nx = %i,  ny = %i\n", nx, ny);
    printf("  nt = %i,  tstop = %g\n", nt, tstop);
    printf("  c = %g\n", c);
    printf("  dtoutput = %g\n", dtoutput);
    printf("  nprocs = %i\n", p2d.numprocs);

    // fill buffers for other MPI processes
    ibuff[0] = nx;
    ibuff[1] = ny;
    ibuff[2] = nt;
    rbuff[0] = tstop;
    rbuff[1] = c;
    rbuff[2] = dtoutput;
  }
  /* broadcast parameters to all procs */
  ierr = MPI_Bcast(ibuff, 3, MPI_INT, 0, MPI_COMM_WORLD);
  check_err(ierr, MPI_COMM_WORLD, "MPI_Bcast");
  ierr = MPI_Bcast(rbuff, 3, MPI_DOUBLE, 0, MPI_COMM_WORLD);
  check_err(ierr, MPI_COMM_WORLD, "MPI_Bcast");
  nx = ibuff[0];
  ny = ibuff[1];
  nt = ibuff[2];
  tstop = rbuff[0];
  c = rbuff[1];
  dtoutput = rbuff[2];
  double ftime = MPI_Wtime();
  double iotime = ftime - stime;

  /* perform parallel decomposition */
  stime = MPI_Wtime();
  p2d.periodic[0] = 1;
  p2d.periodic[1] = 1;
  p2d.pdims[0] = 0;
  p2d.pdims[1] = 0;
  ierr = MPI_Dims_create(p2d.numprocs, 2, p2d.pdims);
  check_err(ierr, MPI_COMM_WORLD, "MPI_Dims_create");

  ierr = MPI_Cart_create(MPI_COMM_WORLD, 2, p2d.pdims, p2d.periodic, 0, &(p2d.comm));
  check_err(ierr, MPI_COMM_WORLD, "MPI_Cart_create");

  ierr = MPI_Cart_coords(p2d.comm, p2d.myid, 2, p2d.pcoords);
  check_err(ierr, p2d.comm, "MPI_Cart_get");

  int is = p2d.pcoords[0]*nx/p2d.pdims[0];
  int ie = (p2d.pcoords[0]+1)*nx/p2d.pdims[0]-1;
  int js = p2d.pcoords[1]*ny/p2d.pdims[1];
  int je = (p2d.pcoords[1]+1)*ny/p2d.pdims[1]-1;

  /* allocate arrays */
  p2d.nxloc = ie-is+1;
  p2d.nyloc = je-js+1;
  double *u  = (double *) malloc((p2d.nxloc*p2d.nyloc) * sizeof(double));
  double *v1 = (double *) malloc((p2d.nxloc*p2d.nyloc) * sizeof(double));
  double *v2 = (double *) malloc((p2d.nxloc*p2d.nyloc) * sizeof(double));
  double *v3 = (double *) malloc((p2d.nxloc*p2d.nyloc) * sizeof(double));

  /* set grid spacing */
  double dx = 1.0/nx;
  double dy = 1.0/ny;

  /* set initial conditions */
  initialize(u, v1, v2, v3, c, dx, dy, is, ie, js, je);
  ftime = MPI_Wtime();
  double inittime = ftime - stime;

  /* set initial time, output initial solution */
  double t = 0.0;
  double toutput = 0.0;
  int noutput = 0;
  if (p2d.myid == 0)
    printf("writing output file %i, step = %i, t = %g\n", noutput, 0, t);
  stime = MPI_Wtime();
  output(u, t, nx, ny, noutput, &p2d);
  ftime = MPI_Wtime();
  iotime += ftime - stime;

  /* set time step */
  double dt = (dx < dy) ? dx/c/50.0 : dy/c/50.0;

  /* start time stepping */
  double runtime = 0.0;
  double commtime = 0.0;
  for (it=0; it<nt; it++) {

    /* communicate to get v2 & v3 neighbor values */
    stime = MPI_Wtime();
    ierr = Communication1(v2, v3, &p2d);
    check_err(ierr, p2d.comm, "Communication1");
    ftime = MPI_Wtime();
    commtime += ftime - stime;

    /* start timer */
    stime = MPI_Wtime();

    /* first update v1 to get to half time step
       get relevant values for this location */
    for (j=0; j<p2d.nyloc; j++) {
      for (i=0; i<p2d.nxloc; i++) {

	/* access relevant components of v2 and v3 */
	v2_E = (i == p2d.nxloc-1) ? p2d.v2recvE[j] : v2[idx(i+1,j,p2d.nxloc)];
	v2_W = v2[idx(i,j,p2d.nxloc)];
	v3_N = (j == p2d.nyloc-1) ? p2d.v3recvN[i] : v3[idx(i,j+1,p2d.nxloc)];
	v3_S = v3[idx(i,j,p2d.nxloc)];

	/* update v1*/
	v1[idx(i,j,p2d.nxloc)] += c*dt/dx*(v2_E - v2_W)
                                + c*dt/dy*(v3_N - v3_S);

      } /* for i */
    } /* for j */
    ftime = MPI_Wtime();
    runtime += ftime - stime;

    /* communicate to get v1 neighbor values */
    stime = MPI_Wtime();
    ierr = Communication2(v1, &p2d);
    check_err(ierr, p2d.comm, "Communication2");
    ftime = MPI_Wtime();
    commtime += ftime - stime;

    /* next update v2 & v3 to get to full time step
       get relevant values for this location */
    stime = MPI_Wtime();
    for (j=0; j<p2d.nyloc; j++) {
      for (i=0; i<p2d.nxloc; i++) {

	/* access relevant components of v1 */
	v1_W = (i == 0) ? p2d.v1recvW[j] : v1[idx(i-1,j,p2d.nxloc)];
	v1_E = v1[idx(i,j,p2d.nxloc)];
	v1_S = (j == 0) ? p2d.v1recvS[i] : v1[idx(i,j-1,p2d.nxloc)];
	v1_N = v1[idx(i,j,p2d.nxloc)];

	/* update v2 and v3 */
	v2[idx(i,j,p2d.nxloc)] += c*dt/dx*(v1_E - v1_W);
	v3[idx(i,j,p2d.nxloc)] += c*dt/dy*(v1_N - v1_S);
	
      } /* for i */
    } /* for j */

    /* update solution for plotting */
    for (j=0; j<p2d.nyloc; j++) 
      for (i=0; i<p2d.nxloc; i++) 
	u[idx(i,j,p2d.nxloc)] += dt*v1[idx(i,j,p2d.nxloc)];

    /* update time */
    t = t + dt;
    ftime = MPI_Wtime();
    runtime += ftime - stime;

    /* stop simulation if we've reached tstop */
    if (t >= tstop)  break;

    /* output solution periodically */
    if ( t - (toutput+dtoutput) > -1.e-14) {
      stime = MPI_Wtime();
      toutput = t;
      noutput++;
      if (p2d.myid == 0)
	printf("writing output file %i, step = %i, t = %g\n", noutput, it, t);
      output(u, t, nx, ny, noutput, &p2d);
      ftime = MPI_Wtime();
      iotime += ftime - stime;
    }

  } /* for it */
  
  /* output final solution */
  stime = MPI_Wtime();
  toutput = t;
  noutput++;
  if (p2d.myid == 0)
    printf("writing output file %i, step = %i, t = %g\n", noutput, it, t);
  output(u, t, nx, ny, noutput, &p2d);
  ftime = MPI_Wtime();
  iotime += ftime - stime;

  if (p2d.myid == 0) {
    printf(" total initialization time = %.16e\n", inittime);
    printf(" total input/output time   = %.16e\n", iotime);
    printf(" total communication time  = %.16e\n", commtime);
    printf(" total simulation time     = %.16e\n", runtime);
  }

  /* free memory */
  free(u);
  free(v1);
  free(v2);
  free(v3);
  free_parallel_decomp(&p2d);

  /* finalize MPI */
  ierr = MPI_Finalize();

  return 0;
} /* end main */



void create_parallel_decomp(parallel_decomp *p2d) {
  p2d->pdims[0] = 1;
  p2d->pdims[1] = 1;
  p2d->periodic[0] = 1;
  p2d->periodic[1] = 1;
  p2d->pcoords[0] = 0;
  p2d->pcoords[1] = 0;
  p2d->nxloc = 0;
  p2d->nyloc = 0;
  p2d->myid = -1;
  p2d->numprocs = -1;
  p2d->nbE = -1;
  p2d->nbW = -1;
  p2d->nbN = -1;
  p2d->nbS = -1;
  p2d->v2recvE = NULL;
  p2d->v3recvN = NULL;
  p2d->v2sendW = NULL;
  p2d->v3sendS = NULL;
  p2d->v1recvW = NULL;
  p2d->v1recvS = NULL;
  p2d->v1sendE = NULL;
  p2d->v1sendN = NULL;
}



void free_parallel_decomp(parallel_decomp *p2d) {
  free(p2d->v2recvE);
  free(p2d->v3recvN);
  free(p2d->v2sendW);
  free(p2d->v3sendS);
  free(p2d->v1recvW);
  free(p2d->v1recvS);
  free(p2d->v1sendE);
  free(p2d->v1sendN);
}



int Communication1(double *v2, double *v3, parallel_decomp *p2d) {

  /* declarations */
  int i, j, ierr, nbcoords[2];
  
  /* allocate send/recv arrays if NULL */
  if (p2d->v2recvE == NULL)
    p2d->v2recvE = (double *) malloc( p2d->nyloc * sizeof(double) );
  if (p2d->v3recvN == NULL)
    p2d->v3recvN = (double *) malloc( p2d->nxloc * sizeof(double) );
  if (p2d->v2sendW == NULL)
    p2d->v2sendW = (double *) malloc( p2d->nyloc * sizeof(double) );
  if (p2d->v3sendS == NULL)
    p2d->v3sendS = (double *) malloc( p2d->nxloc * sizeof(double) );

  /* initialize send/recv buffers */
  for (j=0; j<p2d->nyloc; j++)  p2d->v2sendW[j] = v2[idx(0,j,p2d->nxloc)];
  for (i=0; i<p2d->nxloc; i++)  p2d->v3sendS[i] = v3[idx(i,0,p2d->nxloc)];
  for (j=0; j<p2d->nyloc; j++)  p2d->v2recvE[j] = 0.0;
  for (i=0; i<p2d->nxloc; i++)  p2d->v3recvN[i] = 0.0;

  /* determine process 'neighbors' */
  if (p2d->nbE < 0) {
    nbcoords[0] = p2d->pcoords[0]-1;
    nbcoords[1] = p2d->pcoords[1];
    ierr = MPI_Cart_rank(p2d->comm, nbcoords, &(p2d->nbW));
    check_err(ierr, p2d->comm, "MPI_Cart_rank");

    nbcoords[0] = p2d->pcoords[0]+1;
    nbcoords[1] = p2d->pcoords[1];
    ierr = MPI_Cart_rank(p2d->comm, nbcoords, &(p2d->nbE));
    check_err(ierr, p2d->comm, "MPI_Cart_rank");

    nbcoords[0] = p2d->pcoords[0];
    nbcoords[1] = p2d->pcoords[1]-1;
    ierr = MPI_Cart_rank(p2d->comm, nbcoords, &(p2d->nbS));
    check_err(ierr, p2d->comm, "MPI_Cart_rank");

    nbcoords[0] = p2d->pcoords[0];
    nbcoords[1] = p2d->pcoords[1]+1;
    ierr = MPI_Cart_rank(p2d->comm, nbcoords, &(p2d->nbN));
    check_err(ierr, p2d->comm, "MPI_Cart_rank");
  }  /* end if (nbE < 0) */

  /* open send/receive channels */
  MPI_Request req[4];
  ierr = MPI_Irecv(p2d->v2recvE, p2d->nyloc, MPI_DOUBLE,
                   p2d->nbE, 2, p2d->comm, &(req[0]));
  check_err(ierr, p2d->comm, "MPI_Irecv");

  ierr = MPI_Irecv(p2d->v3recvN, p2d->nxloc, MPI_DOUBLE,
                   p2d->nbN, 4, p2d->comm, &(req[1]));
  check_err(ierr, p2d->comm, "MPI_Irecv");

  ierr = MPI_Isend(p2d->v2sendW, p2d->nyloc, MPI_DOUBLE,
                   p2d->nbW, 2, p2d->comm, &(req[2]));
  check_err(ierr, p2d->comm, "MPI_Isend");

  ierr = MPI_Isend(p2d->v3sendS, p2d->nxloc, MPI_DOUBLE,
                   p2d->nbS, 4, p2d->comm, &(req[3]));
  check_err(ierr, p2d->comm, "MPI_Isend");

  /* wait until all communications finish */
  MPI_Status stat[4];
  ierr = MPI_Waitall(4, req, stat);
  check_err(ierr, p2d->comm, "MPI_Waitall");

  return 0;
}




int Communication2(double *v1, parallel_decomp *p2d) {
  
  /* declarations */
  int i, j, ierr, nbcoords[2];
  
  /* allocate send/recv arrays if NULL */
  if (p2d->v1recvW == NULL)
    p2d->v1recvW = (double *) malloc( p2d->nyloc * sizeof(double) );
  if (p2d->v1recvS == NULL)
    p2d->v1recvS = (double *) malloc( p2d->nxloc * sizeof(double) );
  if (p2d->v1sendE == NULL)
    p2d->v1sendE = (double *) malloc( p2d->nyloc * sizeof(double) );
  if (p2d->v1sendN == NULL)
    p2d->v1sendN = (double *) malloc( p2d->nxloc * sizeof(double) );

  /* initialize send/recv buffers */
  for (j=0; j<p2d->nyloc; j++)  p2d->v1sendE[j] = v1[idx(p2d->nxloc-1,j,p2d->nxloc)];
  for (i=0; i<p2d->nxloc; i++)  p2d->v1sendN[i] = v1[idx(i,p2d->nyloc-1,p2d->nxloc)];
  for (j=0; j<p2d->nyloc; j++)  p2d->v1recvW[j] = 0.0;
  for (i=0; i<p2d->nxloc; i++)  p2d->v1recvS[i] = 0.0;

  /* open send/receive channels */
  MPI_Request req[4];
  ierr = MPI_Irecv(p2d->v1recvW, p2d->nyloc, MPI_DOUBLE,
                   p2d->nbW, 2, p2d->comm, &(req[0]));
  check_err(ierr, p2d->comm, "MPI_Irecv");

  ierr = MPI_Irecv(p2d->v1recvS, p2d->nxloc, MPI_DOUBLE,
                   p2d->nbS, 4, p2d->comm, &(req[1]));
  check_err(ierr, p2d->comm, "MPI_Irecv");

  ierr = MPI_Isend(p2d->v1sendE, p2d->nyloc, MPI_DOUBLE,
                   p2d->nbE, 2, p2d->comm, &(req[2]));
  check_err(ierr, p2d->comm, "MPI_Isend");

  ierr = MPI_Isend(p2d->v1sendN, p2d->nxloc, MPI_DOUBLE,
                   p2d->nbN, 4, p2d->comm, &(req[3]));
  check_err(ierr, p2d->comm, "MPI_Isend");

  /* wait until all communications finish */
  MPI_Status stat[4];
  ierr = MPI_Waitall(4, req, stat);
  check_err(ierr, p2d->comm, "MPI_Waitall");

  return 0;
}

File: advection/advection_mpi.f90
! Daniel R. Reynolds
! SMU Mathematics
! Math 4370/6370
! 7 February 2015
!-----------------------------------------------------------------
! Description: 
!    Evolves the first-order 2D wave equations in time.
!=================================================================

program advection_mpi
  !===============================================================
  ! inclusions
  use mpi
  use exchange
  implicit none

  ! interfaces
  interface
     subroutine output_mpi(u, t, nxloc, nyloc, pid, pcoords, &
                           nx, ny, px, py, noutput)
       integer, intent(in) :: nxloc, nyloc, pid, pcoords(2)
       integer, intent(in) :: nx, ny, px, py, noutput
       real*8,  intent(in) :: t, u(nxloc,nyloc)
     end subroutine output_mpi
     subroutine initialize_mpi(u,v1,v2,v3,c,dx,dy,xl,yl,nxloc,nyloc,myid)
       integer, intent(in)  :: nxloc, nyloc, myid
       real*8,  intent(in)  :: c, dx, dy, xl, yl
       real*8, dimension(nxloc,nyloc), intent(out) :: u, v1, v2, v3
     end subroutine initialize_mpi
  end interface

  ! declarations
  integer :: nx, ny, nt, j, i, noutput, it, ierr
  integer :: myid, nprocs, ibuff(3), pdims(2), comm
  integer :: pcoords(2), nxloc, nyloc
  real*8  :: tstop, c, dtoutput, dx, dy, y_c, y_h, x_c, x_h, t, toutput, dt
  real*8  :: v2_E, v2_W, v3_N, v3_S, v1_W, v1_E, v1_S, v1_N
  real*8  :: stime, ftime, runtime, iotime, inittime, commtime
  real*8, dimension(:,:), allocatable :: u, v1, v2, v3
  real*8  :: rbuff(3), xl, yl 
  real*8, dimension(:), allocatable :: v2E, v3N, v1W, v1S
  logical :: periods(2)
  namelist /inputs/ nx, ny, nt, tstop, c, dtoutput

  ! internals
  call MPI_Init(ierr)
  if (ierr /= MPI_SUCCESS) then
     write(0,*) 'advection_mpi error in MPI_Init =',ierr
     stop
  endif
  call MPI_Comm_size(MPI_COMM_WORLD, nprocs, ierr)
  if (ierr /= MPI_SUCCESS) then
     write(0,*) 'advection_mpi error in MPI_Comm_size =',ierr
     call MPI_Abort(MPI_COMM_WORLD, 1, ierr)
  endif
  call MPI_Comm_rank(MPI_COMM_WORLD, myid, ierr)
  if (ierr /= MPI_SUCCESS) then
     write(0,*) 'advection_mpi error in MPI_Comm_rank =',ierr
     call MPI_Abort(MPI_COMM_WORLD, 1, ierr)
  endif

  ! initialize timers
  runtime = 0.d0
  iotime = 0.d0
  inittime = 0.d0 
  commtime = 0.d0
  
  stime = MPI_Wtime()
  ! read problem parameters from input file (should be in this order):
  !    nx - number of nodes in x-direction
  !    ny - number of nodes in y-direction
  !    nt - number of time steps
  !    tstop - final time (will stop at nt or stop, whichever is 1st)
  !    c - wave speed
  !    dtoutput - time frequency for outputting solutions
  if (myid == 0) then
     open(100,file='input.txt')
     read(100,inputs)
     close(100)
     ibuff = (/ nx, ny, nt /)
     rbuff = (/ tstop, c, dtoutput /)
  end if
  call MPI_Bcast(ibuff, 3, MPI_INTEGER, 0, MPI_COMM_WORLD, ierr)
  if (ierr /= MPI_SUCCESS) then
     write(0,*) 'advection_mpi error in MPI_Bcast =',ierr
     call MPI_Abort(MPI_COMM_WORLD, 1, ierr)
  endif
  call MPI_Bcast(rbuff, 3, MPI_REAL8, 0, MPI_COMM_WORLD, ierr)
  if (ierr /= MPI_SUCCESS) then
     write(0,*) 'advection_mpi error in MPI_Bcast =',ierr
     call MPI_Abort(MPI_COMM_WORLD, 1, ierr)
  endif
  nx = ibuff(1)
  ny = ibuff(2)
  nt = ibuff(3)
  tstop = rbuff(1)
  c = rbuff(2)
  dtoutput = rbuff(3)
  
  if (myid == 0) then
     print *, '  '
     print *, 'Running wave problem:'
     print *, '   nx =',nx,',  ny =',ny
     print *, '   nt =',nt,',  tstop =',tstop
     print *, '   c =',c
     print *, '   dtoutput =',dtoutput
     print *, '   nprocs =',nprocs
     print *, '  '
  end if
  ftime = MPI_Wtime()
  iotime = iotime + ftime-stime

  stime = MPI_Wtime()
  ! create cartesian communicator
  pdims = (/ 0, 0 /)
  periods = .true.
  call MPI_Dims_create(nprocs, 2, pdims, ierr)
  if (ierr /= MPI_SUCCESS) then
     write(0,*) 'advection_mpi error in MPI_Dims_create =',ierr
     call MPI_Abort(MPI_COMM_WORLD, 1, ierr)
  endif
  call MPI_Cart_create(MPI_COMM_WORLD, 2, pdims, periods, &
       .false., comm, ierr)
  if (ierr /= MPI_SUCCESS) then
     write(0,*) 'advection_mpi error in MPI_Cart_create =',ierr
     call MPI_Abort(MPI_COMM_WORLD, 1, ierr)
  endif
  call MPI_Cart_coords(comm, myid, 2, pcoords, ierr)
  if (ierr /= MPI_SUCCESS) then
     write(0,*) 'advection_mpi error in MPI_Cart_coords =',ierr
     call MPI_Abort(MPI_COMM_WORLD, 1, ierr)
  endif

  ! set grid spacing and mesh points
  dx = 1.d0/nx
  dy = 1.d0/ny
  nxloc = nx/pdims(1)
  nyloc = ny/pdims(2)
  xl = dx*nxloc*pcoords(1)
  yl = dy*nyloc*pcoords(2)
  if (pcoords(1) == pdims(1)-1) then
     if (nxloc*pdims(1) /= nx) then
        nxloc = nx - nxloc*(pdims(1)-1)
     end if
  end if
  if (pcoords(2) == pdims(2)-1) then
     if (nyloc*pdims(2) /= ny) then
        nyloc = ny - nyloc*(pdims(2)-1)
     end if
  end if

  ! allocate arrays
  allocate(u(nxloc,nyloc), v1(nxloc,nyloc), &
           v2(nxloc,nyloc), v3(nxloc,nyloc))
  allocate(v2E(nyloc), v3N(nxloc), v1W(nyloc), v1S(nxloc))

  ! set initial conditions
  call initialize_mpi(u,v1,v2,v3,c,dx,dy,xl,yl,nxloc,nyloc,myid)
  ftime = MPI_Wtime()
  inittime = inittime + ftime-stime

  ! set initial time, output initial solution
  !   root outputs problem/parallelism information
  stime = MPI_Wtime()
  t = 0.d0
  toutput = 0.d0
  noutput = 0
  if (myid == 0) print '(A,i4,A,i8,A,es9.2)', &
       'writing output file ',noutput, ', step = ',0,',  t =',t
  call output_mpi(u, t, nxloc, nyloc, myid, pcoords, nx, ny, pdims(1), pdims(2), noutput)
  ftime = MPI_Wtime()
  iotime = iotime + ftime-stime

  ! initialize exchange module
  stime = MPI_Wtime()
  call init_exchange(comm, nxloc, nyloc, pcoords)
  ftime = MPI_Wtime()
  inittime = inittime + ftime-stime

  ! set time step
  dt = min(dx/c/5.d1, dy/c/5.d1)

  ! start time stepping
  do it=1,nt

     ! communicate to get v2 and v3 neighbor values
     stime = MPI_Wtime()
     call do_exchange2(v2, v3, v2E, v3N, ierr)
     if (ierr /= MPI_SUCCESS) then
        write(0,*) 'advection_mpi error in do_exchange2 =',ierr
        call MPI_Abort(MPI_COMM_WORLD, 1, ierr)
     endif
     ftime = MPI_Wtime()
     commtime = commtime + ftime-stime

     stime = MPI_Wtime()
     ! first update v1 to get to half time step
     !    get relevant values for this location
     do j=1,nyloc
        do i=1,nxloc

           ! access relevant components of v2 and v3
           if (i==nxloc) then
              v2_E = v2E(j)
           else
              v2_E = v2(i+1,j)
           end if
           v2_W = v2(i,j)
           if (j==nyloc) then
              v3_N = v3N(i)
           else
              v3_N = v3(i,j+1)
           end if
           v3_S = v3(i,j)

           ! update v1
           v1(i,j) = v1(i,j) + c*dt/dx*(v2_E - v2_W) + c*dt/dy*(v3_N - v3_S)

        end do
     end do
     ftime = MPI_Wtime()
     runtime = runtime + ftime-stime

     ! communicate to get v1 neighbor values
     stime = MPI_Wtime()
     call do_exchange1(v1, v1W, v1S, ierr)
     if (ierr /= MPI_SUCCESS) then
        write(0,*) 'advection_mpi error in do_exchange1 =',ierr
        call MPI_Abort(MPI_COMM_WORLD, 1, ierr)
     endif
     ftime = MPI_Wtime()
     commtime = commtime + ftime-stime

     stime = MPI_Wtime()
     ! next update v2 & v3 to get to full time step
     !    get relevant values for this location
     do j=1,nyloc
        do i=1,nxloc

           ! access relevant components of v1
           if (i==1) then
              v1_W = v1W(j)
           else
              v1_W = v1(i-1,j)
           end if
           v1_E = v1(i,j)
           if (j==1) then
              v1_S = v1S(i)
           else
              v1_S = v1(i,j-1)
           end if
           v1_N = v1(i,j)

           ! update v2 and v3
           v2(i,j) = v2(i,j) + c*dt/dx*(v1_E - v1_W)
           v3(i,j) = v3(i,j) + c*dt/dy*(v1_N - v1_S)

        end do
     end do

     ! update solution
     u = u + dt*v1

     ! update time
     t = t + dt
     ftime = MPI_Wtime()
     runtime = runtime + ftime-stime

     ! stop simulation if we've reached tstop
     if (t >= tstop)  exit

     ! output solution periodically
     if (abs(t-toutput-dtoutput) <= 1.e-14) then
        stime = MPI_Wtime()
        toutput = t
        noutput = noutput+1
        if (myid == 0) print '(A,i4,A,i8,A,es9.2)', &
             'writing output file ',noutput,', step = ',it,',  t =',t
        call output_mpi(u, t, nxloc, nyloc, myid, pcoords, nx, ny, pdims(1), pdims(2), noutput)
        ftime = MPI_Wtime()
        iotime = iotime + ftime-stime
     end if

  end do


  ! output final solution 
  stime = MPI_Wtime()
  toutput = t
  noutput = noutput+1
  if (myid == 0) print '(A,i4,A,i8,A,es9.2)', &
       'writing output file ',noutput,', step = ',it,',  t =',t
  call output_mpi(u, t, nxloc, nyloc, myid, pcoords, nx, ny, pdims(1), pdims(2), noutput)
  ftime = MPI_Wtime()
  iotime = iotime + ftime-stime
  
  ! output runtime
  if (myid == 0) then
     print *, ' total initialization time = ', inittime
     print *, ' total input/output time   = ', iotime
     print *, ' total communication time  = ', commtime
     print *, ' total simulation time     = ', runtime
  endif

  ! free memory
  call free_exchange()
  deallocate(u, v1, v2, v3)
  deallocate(v2E, v3N, v1W, v1S)

  call MPI_Finalize(ierr)

  ! end program
end program advection_mpi
!=================================================================
